source/extensions/filters/http/geoip/geoip_filter.cc

```diff
@@ -0,0 +1,132 @@
+    geo_header = configured_geo_header;
+    stat_name_set_->rememberBuiltin(absl::StrCat(configured_geo_header, ".hit"));
+    stat_name_set_->rememberBuiltin(absl::StrCat(configured_geo_header, ".total"));
+    at_least_one_geo_header_configured_ = true;
+  }
+}
+
+void GeoipFilterConfig::incCounter(Stats::StatName name) {
+  Stats::SymbolTable::StoragePtr storage = scope_.symbolTable().join({stats_prefix_, name});
+  scope_.counterFromStatName(Stats::StatName(storage.get())).inc();
+}
+
+GeoipFilter::GeoipFilter(GeoipFilterConfigSharedPtr config, DriverSharedPtr driver)
+    : config_(config), driver_(std::move(driver)) {}
+
+GeoipFilter::~GeoipFilter() = default;
+
+void GeoipFilter::onDestroy() {}
+
+Http::FilterHeadersStatus GeoipFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  Network::Address::InstanceConstSharedPtr remote_address;
+  if (config_->use_xff() && config_->xffNumTrustedHops() > 0) {
+    remote_address =
+        Envoy::Http::Utility::getLastAddressFromXFF(headers, config_->xffNumTrustedHops()).address_;
+  }
+  // If `config_->use_xff() == false` or xff header has not been populated for some reason.
+  if (!remote_address) {
+    remote_address = callbacks_->streamInfo().downstreamAddressProvider().remoteAddress();
+  }
+  lookupGeolocationData(config_->geoCountryHeader(), remote_address, &Driver::getCountry, headers);
+  lookupGeolocationData(config_->geoCityHeader(), remote_address, &Driver::getCity, headers);
+  lookupGeolocationData(config_->geoRegionHeader(), remote_address, &Driver::getRegion, headers);
+  lookupGeolocationData(config_->geoAsnHeader(), remote_address, &Driver::getAsn, headers);
+  // todo(nezdolik) switch to batch lookup for anon data. Such lookup is supported by maxmind lib.
+  lookupGeolocationAnonData(config_->geoAnonHeader(), remote_address, &Driver::getIsAnonymous,
+                            headers);
+  lookupGeolocationAnonData(config_->geoAnonVpnHeader(), remote_address, &Driver::getIsAnonymousVpn,
+                            headers);
+  lookupGeolocationAnonData(config_->geoAnonHostingHeader(), remote_address,
+                            &Driver::getIsAnonymousHostingProvider, headers);
+  lookupGeolocationAnonData(config_->geoAnonTorHeader(), remote_address,
+                            &Driver::getIsAnonymousTorExitNode, headers);
+  lookupGeolocationAnonData(config_->geoAnonProxyHeader(), remote_address,
+                            &Driver::getIsAnonymousPublicProxy, headers);
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+void GeoipFilter::lookupGeolocationData(
+    const absl::optional<std::string>& geo_header,
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    const absl::optional<std::string>& (Driver::*func)(
+        const Network::Address::InstanceConstSharedPtr& address) const,
+    Http::RequestHeaderMap& headers) {
+  if (geo_header) {
+    auto lookupResult = ((*driver_).*func)(remote_address);
```

Comment:
Are we confident that this is going to be fast enough to not leave the worker thread blocked on I/O for any significant amount of time? It seems likely that this would be risky to extend to other geoip sources for that reason. Might it be better to implement with callbacks and handle asynchronously?

Reply:
Agree, async fits much better for this use case